File: tests/benchmarks/coro/sleep_storm/sleep_storm.erl
%% Sleep storm benchmark - Erlang equivalent
%% N processes sleep simultaneously, measure scheduling overhead
%% Usage: erl -noshell -s sleep_storm main -s init stop
-module(sleep_storm).
-export([main/0, main/2]).

main() ->
    main(10000, 100).

main(N, SleepMs) ->
    io:format("=== Sleep Storm 测试 (Erlang) ===~n"),
    io:format("协程数量: ~p~n", [N]),
    io:format("Sleep时间: ~p ms~n", [SleepMs]),

    Self = self(),
    T0 = erlang:monotonic_time(millisecond),

    [spawn(fun() ->
        Before = erlang:monotonic_time(millisecond),
        timer:sleep(SleepMs),
        After = erlang:monotonic_time(millisecond),
        Self ! {done, After - Before}
    end) || _ <- lists:seq(1, N)],

    SpawnTime = erlang:monotonic_time(millisecond) - T0,
    io:format("创建时间: ~p ms~n", [SpawnTime]),

    {TotalSleep, MinSleep, MaxSleep} = collect_sleep(N, 0, 999999, 0),

    Elapsed = erlang:monotonic_time(millisecond) - T0,
    io:format("总时间: ~p ms~n", [Elapsed]),
    io:format("平均sleep: ~p ms~n", [TotalSleep div N]),
    io:format("最小sleep: ~p ms~n", [MinSleep]),
    io:format("最大sleep: ~p ms~n", [MaxSleep]),
    io:format("调度开销: ~p ms~n", [MaxSleep - SleepMs]).

collect_sleep(0, Total, Min, Max) ->
    {Total, Min, Max};
collect_sleep(N, Total, Min, Max) ->
    receive {done, Actual} ->
        NewMin = case Actual < Min of true -> Actual; false -> Min end,
        NewMax = case Actual > Max of true -> Actual; false -> Max end,
        collect_sleep(N - 1, Total + Actual, NewMin, NewMax)
    end.

File: tests/benchmarks/coro/sleep_storm/sleep_storm.go
// 作者：xingleixu@gmail.com
// Sleep Storm 测试
// 大量 goroutine 同时 sleep 后唤醒，测试定时器和调度效率

package main

import (
	"fmt"
	"os"
	"strconv"
	"sync"
	"time"
)

func main() {
	N := 10000      // 协程数量
	SLEEP_MS := 100 // sleep 时间（毫秒）

	if len(os.Args) > 1 {
		if n, err := strconv.Atoi(os.Args[1]); err == nil {
			N = n
		}
	}
	if len(os.Args) > 2 {
		if s, err := strconv.Atoi(os.Args[2]); err == nil {
			SLEEP_MS = s
		}
	}

	fmt.Println("=== Sleep Storm 测试 ===")
	fmt.Println("协程数量:", N)
	fmt.Println("Sleep时间:", SLEEP_MS, "ms")

	sleepDuration := time.Duration(SLEEP_MS) * time.Millisecond
	results := make([]float64, N)
	var wg sync.WaitGroup

	start := time.Now()

	// 同时启动所有 goroutine
	for i := 0; i < N; i++ {
		wg.Add(1)
		go func(idx int) {
			defer wg.Done()
			before := time.Now()
			time.Sleep(sleepDuration)
			after := time.Now()
			results[idx] = float64(after.Sub(before)) / float64(time.Millisecond)
		}(i)
	}

	spawnTime := time.Since(start)
	spawnMs := float64(spawnTime) / float64(time.Millisecond)
	fmt.Printf("创建时间: %.3f ms\n", spawnMs)

	// 等待所有协程完成
	wg.Wait()

	elapsed := time.Since(start)
	elapsedMs := float64(elapsed) / float64(time.Millisecond)

	// 统计
	var totalSleep float64
	var maxSleep float64 = 0
	var minSleep float64 = 999999

	for _, actual := range results {
		totalSleep += actual
		if actual > maxSleep {
			maxSleep = actual
		}
		if actual < minSleep {
			minSleep = actual
		}
	}

	fmt.Printf("总时间: %.3f ms\n", elapsedMs)
	fmt.Printf("平均sleep: %.3f ms\n", totalSleep/float64(N))
	fmt.Printf("最小sleep: %.3f ms\n", minSleep)
	fmt.Printf("最大sleep: %.3f ms\n", maxSleep)
	fmt.Printf("调度开销: %.3f ms\n", maxSleep-float64(SLEEP_MS))
}
